src/gfx/pal_packing.cpp

```diff
@@ -0,0 +1,359 @@
+		Iota operator++() {
+			++i;
+			return *this;
+		}
+		Iota operator++(int) {
+			Iota copy = *this;
+			++i;
+			return copy;
+		}
+	};
+
+	// Begin with all proto-palettes queued up for insertion
+	std::queue queue(std::deque<ProtoPalAttrs>(Iota{0}, Iota{protoPalettes.size()}));
+	// Begin with no pages
+	std::vector<AssignedProtos> assignments{};
+
+	while (!queue.empty()) {
+		ProtoPalAttrs const &attrs = queue.front();
+		ProtoPalette const &protoPal = protoPalettes[attrs.palIndex];
+		size_t bestPalIndex = assignments.size();
+		// We're looking for a palette where the proto-palette's relative size is less than
+		// its actual size; so only overwrite the "not found" index on meeting that criterion
+		double bestRelSize = protoPal.size();
+
+		for (size_t i = 0; i < assignments.size(); ++i) {
+			// Skip the page if this one is banned from it
+			if (attrs.isBannedFrom(i)) {
+				continue;
+			}
+
+			options.verbosePrint("%zu: Rel size: %f (size = %zu)\n", i,
+			                     assignments[i].relSizeOf(protoPal), protoPal.size());
+			if (assignments[i].relSizeOf(protoPal) < bestRelSize) {
+				bestPalIndex = i;
+			}
+		}
+
+		if (bestPalIndex == assignments.size()) {
+			// Found nowhere to put it, create a new page containing just that one
+			assignments.emplace_back(protoPalettes, std::move(attrs));
+		} else {
+			auto &bestPal = assignments[bestPalIndex];
+			// Add the color to that palette
+			bestPal.assign(std::move(attrs));
+
+			// If this overloads the palette, get it back to normal (if possible)
+			while (bestPal.volume() > 4) {
+				options.verbosePrint("Palette %zu is overloaded! (%zu > 4)\n", bestPalIndex,
+				                     bestPal.volume());
+
+				// Look for a proto-pal minimizing "efficiency" (size / rel_size)
+				auto efficiency = [&bestPal](ProtoPalette const &pal) {
+					return pal.size() / bestPal.relSizeOf(pal);
+				};
+				auto [minEfficiencyIter, maxEfficiencyIter] =
+					std::minmax_element(bestPal.begin(), bestPal.end(),
+				                        [&efficiency, &protoPalettes](ProtoPalAttrs const &lhs,
+				                                                      ProtoPalAttrs const &rhs) {
+											return efficiency(protoPalettes[lhs.palIndex])
+					                               < efficiency(protoPalettes[rhs.palIndex]);
+										});
+
+				// All efficiencies are identical iff min equals max
+				// TODO: maybe not ideal to re-compute these two?
+				// TODO: yikes for float comparison! I *think* this threshold is OK?
+				if (efficiency(protoPalettes[maxEfficiencyIter->palIndex])
+				        - efficiency(protoPalettes[minEfficiencyIter->palIndex])
+				    < .001) {
+					break;
+				}
+
+				// Remove the proto-pal with minimal efficiency
+				queue.emplace(std::move(*minEfficiencyIter));
+				queue.back().banFrom(bestPalIndex); // Ban it from this palette
+				bestPal.remove(minEfficiencyIter);
+			}
+		}
+
+		queue.pop(); // DO NOT USE `attrs` BEYOND THIS POINT
```

Comment:
`attrs` is defined in this scope of this loop anyway, so that's not a concern. If it were I'd suggest a plain `{ }` block around the scope in which it's valid.

Reply:
The comment is there in case code is added after this `pop` later on.

Reply:
How about changing the `while` to `for (std::queue queue(std::deque<ProtoPalAttrs>(Iota{0}, Iota{protoPalettes.size()})); !queue.empty(); queue.pop())`? IMO this groups the related `queue` operations together well, and scopes `attrs` prior to the `queue.pop()`.

Reply:
Okay. Hopefully we won't need to add code after the `queue.pop()` later on, then.